stepflow-rs/crates/stepflow-execution/src/flow_executor.rs

```diff
@@ -0,0 +1,871 @@
+use stepflow_plugin::ExecutionContext;
+use stepflow_state::StateStore;
+use uuid::Uuid;
+
+use crate::scheduler::{Scheduler, SchedulerDecision};
+use crate::state::ItemsState;
+use crate::step::execute_step_async;
+use crate::task::{Task, TaskResult};
+use crate::{ExecutionError, Result};
+
+/// Executor for running workflows with one or more items.
+///
+/// FlowExecutor coordinates execution across multiple items, using a scheduler
+/// to determine task ordering and respecting concurrency limits. It supports
+/// both run-to-completion and streaming execution modes.
+///
+/// # Example
+///
+/// ```ignore
+/// let flow = Arc::new(flow);
+/// let inputs = vec![input1, input2, input3];
+/// let state = ItemsState::batch(flow.clone(), inputs, variables);
+///
+/// let mut executor = FlowExecutor::new(
+///     stepflow.clone(),
+///     run_id,
+///     flow_id,
+///     state,
+///     Box::new(DepthFirstScheduler::new()),
+///     10, // max concurrency
+///     state_store,
+/// );
+///
+/// let results = executor.execute_to_completion().await?;
+/// ```
+pub struct FlowExecutor {
+    /// Reference to the StepflowExecutor for plugin access.
+    stepflow: Arc<crate::StepflowExecutor>,
+    /// Unique run ID for this execution.
+    run_id: Uuid,
+    /// Flow ID (content hash) for the workflow.
+    flow_id: BlobId,
+    /// Per-item execution state.
+    state: ItemsState,
+    /// Scheduler for task ordering.
+    scheduler: Box<dyn Scheduler>,
+    /// Maximum concurrent tasks.
+    max_in_flight: usize,
+    /// State store for persisting results.
+    state_store: Arc<dyn StateStore>,
+}
+
+impl FlowExecutor {
+    /// Create a new items executor.
+    ///
+    /// # Arguments
+    ///
+    /// * `stepflow` - Reference to the StepflowExecutor for plugin access
+    /// * `run_id` - Unique run ID for this execution
+    /// * `flow_id` - Flow ID (content hash) for the workflow
+    /// * `state` - Pre-initialized ItemsState with items and variables
+    /// * `scheduler` - Scheduler for task ordering
+    /// * `max_in_flight` - Maximum concurrent tasks
+    /// * `state_store` - State store for persisting results
+    pub fn new(
+        stepflow: Arc<crate::StepflowExecutor>,
+        run_id: Uuid,
+        flow_id: BlobId,
+        state: ItemsState,
+        scheduler: Box<dyn Scheduler>,
+        max_in_flight: usize,
+        state_store: Arc<dyn StateStore>,
+    ) -> Self {
+        Self {
+            stepflow,
+            run_id,
+            flow_id,
+            state,
+            scheduler,
+            max_in_flight,
+            state_store,
+        }
+    }
+
+    /// Execute all items to completion.
+    ///
+    /// Returns a vector of FlowResults, one per item, in item index order.
+    /// Results are recorded to the state store as items complete.
+    ///
+    /// Tasks are executed concurrently up to `max_in_flight` at a time.
+    pub async fn execute_to_completion(&mut self) -> Result<Vec<FlowResult>> {
```

Comment:
We could allow this to be a stream of results and include the`ItemIndex` in the result. Then, we wouldn't need the order to match, nor would we need to collect all the results.